packages/flutter/lib/src/material/focus_highlight.dart

```diff
@@ -0,0 +1,212 @@
+/// has focus, and [unfocusedDecoration] when it is not.
+///
+/// By default, it uses the values derived from [Theme.focusHighlightTheme].
+class FocusHighlight extends StatefulWidget {
+  /// Creates a widget that manages a [FocusNode]
+  ///
+  /// The [child] argument is required and must not be null.
+  ///
+  /// The [autofocus], and [showDecorations] arguments must not be null.
+  const FocusHighlight({
+    Key key,
+    @required this.child,
+    this.focusedDecoration,
+    this.unfocusedDecoration,
+    this.duration,
+    this.curve,
+  })  : assert(child != null),
+        super(key: key);
+
+  /// The child widget of this [FocusHighlight].
+  ///
+  /// {@macro flutter.widgets.child}
+  final Widget child;
+
+  /// Defines the decoration to be drawn around the [child] when this widget
+  /// receives the focus.
+  ///
+  /// This defaults to the [FocusHighlightThemeData.focusedDecoration] provided by
+  /// the ambient [Theme] in the context.
+  final Decoration focusedDecoration;
+
+  /// Defines the decoration to be drawn around the [child] when this widget
+  /// does not have the focus.
+  ///
+  /// This defaults to the [FocusHighlightThemeData.unfocusedDecoration] provided by
+  /// the ambient [Theme] in the context.
+  final Decoration unfocusedDecoration;
+
+  /// The duration over which to animate the parameters of the focus highlight
+  /// [decoration].
+  final Duration duration;
+
+  /// The curve to apply when animating the parameters of the focus highlight
+  /// [decoration].
+  final Curve curve;
+
+  @override
+  _FocusHighlightState createState() => _FocusHighlightState();
+}
+
+class _FocusHighlightState extends State<FocusHighlight> {
+  bool _hasFocus;
+  FocusNode _node;
+
+  @override
+  void didUpdateWidget(FocusHighlight oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    final FocusNode newNode = Focus.of(context);
+    if (newNode != _node) {
+      _node.removeListener(_handleFocusChanged);
+      _node = newNode;
+      _node.addListener(_handleFocusChanged);
+    }
+  }
+
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    _node = Focus.of(context);
+    _node.addListener(_handleFocusChanged);
+    _hasFocus = _node.hasFocus;
+  }
+
+  @override
+  void dispose() {
+    _node?.removeListener(_handleFocusChanged);
+    super.dispose();
+  }
+
+  void _handleFocusChanged() {
+    if (_hasFocus != _node.hasFocus) {
+      setState(() {
+        _hasFocus = _node.hasFocus;
+      });
+    }
+  }
+
+  Decoration _getDecoration(BuildContext context) {
+    if (_hasFocus) {
+      return widget.focusedDecoration ?? Theme.of(context).focusHighlightTheme.focusedDecoration;
+    } else {
+      return widget.unfocusedDecoration ?? Theme.of(context).focusHighlightTheme.unfocusedDecoration;
+    }
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return AnimatedContainer(
+      padding: EdgeInsets.zero,
```

Comment:
it's cheaper to not specify at all if it's zero, iirc (unless this is about overriding the decoration? in which case add a comment explaining it so we don't optimize it away) (and make sure it's tested)

Reply:
Removed.